### PR TITLE
[Kwek Kee En] Duke Increments

### PR DESCRIPTION
**Pull request by kwekke**

### CODE REVIEW

File: src/main/java/duke/Duke.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        } catch (DukeException e) {
+            return e.getMessage();
+        }
+    }
+
+    /**
+     * Returns Duke's welcome message.
+     * @return Duke's welcome message in a string
+     */
+    public String getWelcomeMessage() {
```
Comment (evelynejuliet) [inline src/main/java/duke/Duke.java:53]: I like how you update the user on how many task he/she has in the list in the welcome message. It is very well-thought of.

File: src/main/java/command/Command.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package command;
+
+import duke.Storage;
+import duke.TaskList;
+import exception.DukeException;
+
+/**
+ * Represents a Command that Parser returns to the main logic in Duke.
+ */
+public abstract class Command {
```
Comment (evelynejuliet) [inline src/main/java/command/Command.java:13]: I like how you created an abstract class Command to categorize the different commands that Parser returns, makes your code more organized.

**Review by evelynejuliet — COMMENTED** [review]
Overall good job! Your code is very readable and understandable. One thing you can do more is try experimenting with the use of stubs in your JUnit tests to practice and gain more exposure! :)

File: src/main/java/command/DeadlineCommand.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+     * @param tasks TaskList which stores the list of tasks.
+     * @param storage Storage which saves the task into the text file.
+     * @throws DukeException DukeException that may arise from invalid inputs.
+     */
+    public String execute(TaskList tasks, Storage storage) throws DukeException {
+        Task task = new DeadlineTask(description, time);
+        tasks.addTask(task);
+        storage.save(tasks);
+        return ("Got it. I've added this task:\n " + task + "\n" + "Now you have " + tasks.getSize() + " task"
+                + (tasks.getSize() == 1 ? " " : "s ") + "in the list.");
```
Comment (ZhangHuafan) [inline src/main/java/command/DeadlineCommand.java]: I like how you detect whether there should be an "s"

File: src/main/java/command/DeleteCommand.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class DeleteCommand extends Command {
+    private int index;
+
+    /**
+     * Creates a DeleteCommand with a given index to delete from the TaskList.
+     * @param index Task index to be deleted.
+     */
+    public DeleteCommand(int index) {
+        this.index = index - 1;
```
Comment (ZhangHuafan) [inline src/main/java/command/DeleteCommand.java]: Maybe consider initializing DeleteCommand(order - 1) to make it more understandable.

File: src/main/java/command/DeleteCommand.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            throw new DukeException("You have no tasks to delete.");
+        } else if (index < 0 || index >= tasks.getSize()) {
+            throw new DukeInvalidTaskIndexException("delete", tasks.getSize());
+        } else {
+            Task t = tasks.getTask(index);
+            tasks.deleteTask(index);
+            storage.save(tasks);
+            return ("Noted. I've removed this task:\n" + t + "\n"
+                    + "Now you have " + tasks.getSize() + " task"
+                    + ((tasks.getSize() - 1) == 1 ? " " : "s ") + "in the list.");
```
Comment (ZhangHuafan) [inline src/main/java/command/DeleteCommand.java]: May consider moving the printing and formatting part to UI class.

File: src/main/java/command/DoneCommand.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     * @throws DukeException DukeException that may arise from invalid inputs.
+     */
+    public String execute(TaskList tasks, Storage storage) throws DukeException {
+        if (tasks.getSize() == 0) {
+            throw new DukeException("You have no tasks to do.");
+        } else if (index < 0 || index >= tasks.getSize()) {
+            throw new DukeInvalidTaskIndexException("do", tasks.getSize());
+        } else {
+            tasks.doneTask(index);
+            storage.save(tasks);
```
Comment (ZhangHuafan) [inline src/main/java/command/DoneCommand.java:37]: I like how you update the task list whenever there is a change.

File: src/main/java/command/EventCommand.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import exception.DukeException;
+import task.EventTask;
+import task.Task;
+
+/**
+ * Represents a Command which adds an EventTask to the TaskList.
+ */
+public class EventCommand extends Command {
+    private String description;
+    private String time;
```
Comment (ZhangHuafan) [inline src/main/java/command/EventCommand.java]: May consider following the instruction to change time as a LocalDateTime object.

File: src/main/java/command/ListCommand.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    /**
+     * Prompts Ui to print the list of tasks.
+     * @param tasks TaskList which stores the list of tasks.
+     * @param storage Storage which saves the task into the text file.
+     */
+    public String execute(TaskList tasks, Storage storage) {
+        return tasks.toString();
+    }
+
+    public boolean isExit() {
```
Comment (ZhangHuafan) [inline src/main/java/command/ListCommand.java:22]: May consider inheriting isExit() method from the super class so you don't need to write it again in every child classes.

File: src/main/java/duke/Parser.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        case "find":
+            if (token.length <= 1) {
+                throw new DukeException("Give me a goddamn keyword to find.");
+            }
+            return new FindCommand(input.substring(4).trim());
+        default:
+            throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
+}
```
Comment (ZhangHuafan) [inline src/main/java/duke/Parser.java:122]: Comparing how you wrote this part with my code, I just found it is much clearer to use switch-cases for all the input though there may be repeated checkings.

File: src/main/java/duke/Storage.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        try {
+            if (file.exists()) {
+                sc = new Scanner(file);
+            } else {
+                file.createNewFile();
+                sc = new Scanner(file);
+            }
+        } catch (IOException e) {
+            throw new DukeException("IOException from invalid filePath");
+        }
```
Comment (ZhangHuafan) [inline src/main/java/duke/Storage.java:33]: I am not sure but maybe it is better to keep IOException as IOException?

File: src/main/java/duke/Storage.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                time = token[3];
+                task = new DeadlineTask(token[2].trim(), time);
+                break;
+            case "E":
+                time = token[3];
+                task = new EventTask(token[2].trim(), time);
+                break;
+            default:
+                System.out.println(tokenString);
+                throw new DukeException("Corrupted file");
```
Comment (ZhangHuafan) [inline src/main/java/duke/Storage.java]: I like how you take use of default.

File: src/main/java/exception/DukeInvalidTaskTimeException.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package exception;
+
+public class DukeInvalidTaskTimeException extends DukeException {
+
+    public DukeInvalidTaskTimeException(String msg) {
+        super("Did you remember to use" + (msg.equals("deadline") ? " \"/by\"?" : " \"/at\"?"));
+    }
+}
```
Comment (ZhangHuafan) [inline src/main/java/exception/DukeInvalidTaskTimeException.java:8]: I do like the way you write these exceptions.

**Review by ZhangHuafan — COMMENTED** [review]
Your coding logic is very clean, so as the Javadoc. The functionality is complete. I have learned how to write child exception classes from you. It is also very nice to save the file whenever a change is made. However, I am not sure whether it would be better to use UI to handle inputs and outputs. Besides, you may consider adding more tests into your project.

**Pull request closed without merging**